pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+	// We are lagging behind. We need to consume the partition in sections.
+	// We iterate through all the ConsumeInterval intervals, starting from the first one after the last commit until the cycleEnd.
+	//cycleEndStartAt := commitRecTime.Truncate(b.cfg.ConsumeInterval).Add(b.cfg.ConsumeInterval + b.cfg.ConsumeIntervalBuffer)
+	cycleEndStartAt, _ := nextCycleEnd(
+		func() time.Time {
+			return state.CommitRecTimestamp
+		},
+		b.cfg.ConsumeInterval,
+		b.cfg.ConsumeIntervalBuffer,
+	)
+	level.Info(b.logger).Log("msg", "partition is lagging behind", "partition", state.Partition, "lag", state.Lag, "cycle_end_start", cycleEndStartAt, "cycle_end", cycleEnd)
+	for ce := cycleEndStartAt; cycleEnd.Sub(ce) >= 0; ce = ce.Add(b.cfg.ConsumeInterval) {
+		// Instead of looking for the commit metadata for each iteration, we use the data returned by consumePartition in the next iteration.
+		err := b.consumePartitionCycle(ctx, state, ce)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", state.Partition, err)
+		}
+
+		// If adding the ConsumeInterval takes it beyond the cycleEnd, we set it to the cycleEnd to not
+		// exit the loop without consuming until cycleEnd.
+		if ce.Compare(cycleEnd) != 0 && ce.Add(b.cfg.ConsumeInterval).After(cycleEnd) {
+			ce = cycleEnd
+		}
+	}
+
+	return nil
+}
+
+// consumePartition consumes records from the given partition until the cycleEnd timestamp.
+// It modifies state, tracking the consumption progress. If the partition is lagging behind, the caller needs to take care
+// of calling consumePartition in sections.
+func (b *BlockBuilder) consumePartitionCycle(ctx context.Context, state *partitionState, cycleEnd time.Time) (retErr error) {
+	blockEndAt := cycleEnd.Truncate(b.cfg.ConsumeInterval)
+	blockEnd := blockEndAt.UnixMilli()
+
+	var (
+		numBlocks     int
+		compactionDur time.Duration
+	)
+	defer func(t time.Time, startState partitionState) {
+		dur := time.Since(t)
+
+		// Don't propagate any errors derived from the canceled context.
+		if errors.Is(retErr, context.Canceled) {
+			retErr = nil
+		}
+
+		if retErr != nil {
+			level.Error(b.logger).Log("msg", "partition consumption failed", "partition", state.Partition, "dur", dur, "start_lag", startState.Lag, "end_lag", state.Lag, "err", retErr)
+			return
+		}
+
+		b.metrics.processPartitionDuration.WithLabelValues(fmt.Sprintf("%d", state.Partition)).Observe(dur.Seconds())
+		level.Info(b.logger).Log("msg", "done consuming", "partition", state.Partition, "dur", dur,
+			"start_lag", startState.Lag, "end_lag", state.Lag, "cycle_end", cycleEnd,
+			"last_block_end", time.UnixMilli(startState.LastBlockEnd), "curr_block_end", time.UnixMilli(state.LastBlockEnd),
+			"last_seen_till", startState.SeenTillOffset, "curr_seen_till", state.SeenTillOffset,
+			"num_blocks", numBlocks, "compact_and_upload_dur", compactionDur)
+	}(time.Now(), *state)
+
+	builder := NewTSDBBuilder(b.logger, b.cfg.DataDir, b.cfg.BlocksStorage, b.limits)
+	defer builder.Close()
```

Comment:
can you handle this error? there is `runutil.CloseWithErrCapture` which in `defer` uses